drivers/platform/xilinx/spi.c

```diff
@@ -70,116 +157,79 @@ int32_t spi_init(struct spi_desc **desc,
 {
 	int32_t				ret;
 	struct spi_desc			*sdesc;
-	struct xil_spi_desc		*xdesc;
+	struct xil_spi_desc 		*xdesc;
 	struct xil_spi_init_param	*xinit;
+	struct spi_engine_desc 		*edesc;
+	enum xil_spi_type		*spi_type;
 
-	sdesc = (struct spi_desc *)malloc(sizeof(*sdesc));
-	xdesc = (struct xil_spi_desc *)malloc(sizeof(*xdesc));
-	if(!sdesc || !xdesc) {
+	sdesc = malloc(sizeof(*sdesc));
+	if(!sdesc) {
 		free(sdesc);
-		free(xdesc);
 		return FAILURE;
 	}
 
+	/* Get only the first member of the sctructure
+	 * Both structures (spi_engine_init_param and spi_init_param)
+	 * have the fisrt member enum xil_spi_type so in this case
+	 * we can read the first member without casting the pointer to
+	 * a certain structure type.
+	*/
+	spi_type = param->extra;
+
+	if (*spi_type == SPI_ENGINE) {
+		edesc = malloc(sizeof(spi_engine_desc));
+		if(!edesc) {
+			free(edesc);
```

Comment:
free(sdesc). And the next one is not ok either. Check the goto failure pattern here https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/aducm3029/spi.c#L105